### ISSUE TITLE
Questions About Compatibility with Let Over Lambda defmacro!

### ISSUE THREAD

**Issue by noctuid** [mabragor/defmacro-enhance#4]
Reading the section about the differences from the version presented in _Let Over Lambda_, I had few comments/questions.

With regards to the `why in the world did g! appeared here? And is o! still accessible?` comment, it is explained that the reasoning behind this is to allow the user to access x before it was evaluated. `g!x` refers to the evaluated form from the `once-only`, and `o!x` refers to the argument as it was passed in (unevaluated). I think the distinction makes sense since `o!x` appears in the parameter list. I don't know how often this would be useful, but maybe it would be worth keeping. What do you think?

As for using `g!-`, I think I like this style more, but I don't know if I liked that it is forced. It's not too hard to translate the _Let Over Lambda_ code into this package's defmacro!, but compatibility could be nice.

**Comment by mabragor** [comment]
Hi!
Thanks for your comments and interest in the project. I think, you're right about `g!-` vs `g!` -- now
I don't like forced things myself and prefer everything to be as flexible as possible -- so both should be allowed.
As for once-only, I think, in situations where you actually need unevaluated form in some places of the macro and evaluated form in others, you should just use `g!` to have appropriate gensyms.
This way, `o!-` parameter of the macro tells the user of the macro: "this parameter is used, as if macro was actually a function". I think, this convention is a useful one to have -- more important than compatibility with the book.

Generally, what I found about defmacro-enhance is that (though at first I thought I would use it everywhere),
it is actually great only at the prototyping stage of development. It allows to write complicated macros quickly and reliably, without keeping all the common bugs in mind. However, explicit use of code-walker makes it implementation-dependent (defmacro!-forms are slow to compile) -- so once the general structure of the project is clear, I rewrite macros using `alexandria:with-gensyms` and `alexandria:once-only`. More exotic features of defmacro-enhance I actually use *really* seldom.

**Comment by noctuid** [comment]
> This way, o!- parameter of the macro tells the user of the macro: "this parameter is used, as if macro was actually a function". 

Yeah, I think you're right. Using just o!- is clearer as well without any extra explanation needed.

> makes it implementation-dependent (defmacro!-forms are slow to compile)

I was also thinking about trying to use `defmacro!` everywhere. Is the compilation time really that significant? I guess I still need to play around with it. It sounds like quicker prototyping would be the main use case as `with-gensyms` and `once-only` are still pretty convenient to use. Thanks for the info.

**Comment by mabragor** [comment]
Usually my workflow is as follows:
  * I have defmacro-enhance in :depends-on and in :use of the package --
this gives no overhead
  * I start writing any macro with just defmacro
  * when I feel that I need some extra features, I just switch to defmacro!
-- this way only a handful of macros end up being defmacro!s
  * much-much later (usually -- half a year) I prune defmacro-enhance (and
some other) temporary dependencies

2016-04-19 6:10 GMT+02:00 Fox Kiester <notifications@github.com>:
